File: tests/scripts/diagnose_linkedin_scraper.py
import requests
import logging
from bs4 import BeautifulSoup

# Setup basic logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("LinkedInDiagnose")


def diagnose_url(url):
    """Diagnose url."""
    print(f"\n🔍 Diagnosing URL: {url}")
    print("-" * 50)

    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        "Sec-Ch-Ua": '"Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"',
        "Sec-Ch-Ua-Mobile": "?0",
        "Sec-Ch-Ua-Platform": '"Windows"',
    }

    try:
        response = requests.get(url, headers=headers, timeout=15)
        print(f"📡 HTTP Status: {response.status_code}")
        print(f"📏 Response Length: {len(response.text)} bytes")
        if response.status_code == 200:
            soup = BeautifulSoup(response.text, "html.parser")
            title = soup.title.string if soup.title else "No title"
            print(f"📑 Page Title: {title}")
            # Check for common bot detection patterns
            if "authwall" in response.url or "security-check" in response.text.lower():
                print("⚠️  DETECTED: Redirected to Authwall/Security check.")

            text = soup.get_text()
            print(f"📝 Extracted Text Preview (first 200 chars):\n{text.strip()[:200]}...")
            if len(text.strip()) < 100:
                print("❌ ERROR: Very little text extracted. JavaScript rendering might be required.")
        else:
            print(f"❌ ERROR: Received non-200 status code.")
            if response.status_code == 403:
                print("💡 Suggestion: Site is blocking basic requests (Forbidden).")
            elif response.status_code == 999:
                print("💡 Suggestion: LinkedIn specific request rejection code.")

    except Exception:
        print(f"❌ EXCEPTION: {e}")
        pass


if __name__ == "__main__":
    url = "https://www.linkedin.com/in/georgehaber/"
    diagnose_url(url)

File: rust_output/src/diagnose_linkedin_scraper.rs
use anyhow::{Result, Context};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Diagnose url.
pub fn diagnose_url(url: String) -> Result<()> {
    // Diagnose url.
    println!("\n🔍 Diagnosing URL: {}", url);
    println!("{}", ("-".to_string() * 50));
    let mut headers = HashMap::from([("User-Agent".to_string(), "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36".to_string()), ("Accept".to_string(), "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8".to_string()), ("Accept-Language".to_string(), "en-US,en;q=0.9".to_string()), ("Sec-Ch-Ua".to_string(), "\"Not_A Brand\";v=\"8\", \"Chromium\";v=\"120\", \"Google Chrome\";v=\"120\"".to_string()), ("Sec-Ch-Ua-Mobile".to_string(), "?0".to_string()), ("Sec-Ch-Ua-Platform".to_string(), "\"Windows\"".to_string())]);
    // try:
    {
        let mut response = /* reqwest::get( */&url).cloned().unwrap_or(/* headers= */ headers);
        println!("📡 HTTP Status: {}", response.status_code);
        println!("📏 Response Length: {} bytes", response.text.len());
        if response.status_code == 200 {
            let mut soup = BeautifulSoup(response.text, "html.parser".to_string());
            let mut title = if soup.title { soup.title.string } else { "No title".to_string() };
            println!("📑 Page Title: {}", title);
            if (response.url.contains(&"authwall".to_string()) || response.text.to_lowercase().contains(&"security-check".to_string())) {
                println!("{}", "⚠️  DETECTED: Redirected to Authwall/Security check.".to_string());
            }
            let mut text = soup.get_text();
            println!("📝 Extracted Text Preview (first 200 chars):\n{}...", text.trim().to_string()[..200]);
            if text.trim().to_string().len() < 100 {
                println!("{}", "❌ ERROR: Very little text extracted. JavaScript rendering might be required.".to_string());
            }
        } else {
            println!("❌ ERROR: Received non-200 status code.");
            if response.status_code == 403 {
                println!("{}", "💡 Suggestion: Site is blocking basic requests (Forbidden).".to_string());
            } else if response.status_code == 999 {
                println!("{}", "💡 Suggestion: LinkedIn specific request rejection code.".to_string());
            }
        }
    }
    // except Exception as _e:
}
